### PR TITLE
Feature/tmb7852/three nodes testing

### PR DESCRIPTION
**Pull request by tmb7852**
Increased number of TPDOs in BMS node to 8. Also, made BMS capable of determining highest and lowest voltages in the battery cells along with the IDs of those cells. This will only work when the BMS is capable of changing from the start state. While in the start state the values will not be accurate.

### CODE REVIEW

File: include/BMS/dev/BQ76952.hpp [hunk 1]
```diff
@@ -159,9 +159,18 @@ class BQ76952 {
...
+     * @param[out] maxVoltage Voltage value of the highes cell voltage of all cells in the pack.
+     *
+     * @param[out] minCellVoltageID The ID of the cell with the lowest voltage.
+     *
+     * @param[out] maxCellVoltageId The ID of the cell with the highest voltage.
+     *
      * @return The status of the read attempt
      */
-    Status getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t* sum);
+    Status getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t* sum, int16_t* minVoltage, uint8_t* minCellVoltageID, int16_t* maxVoltage, uint8_t* maxCellVoltageId);
```
Comment (cbolles) [inline include/BMS/dev/BQ76952.hpp:173]: NIT: May want to make a struct to hold voltage reading parameters to wrap up all the data and reduce the number of parameters

File: src/BMS.cpp [hunk 2]
```diff
@@ -109,15 +111,15 @@ void BMS::initializationErrorState() {
     if (stateChanged) {
         bmsOK.writePin(BMS_NOT_OK);
         stateChanged = false;
-        clearVoltageReadings();
+        //clearVoltageReadings();
```
Comment (cbolles) [inline src/BMS.cpp:114]: NIT: Shouldn't leave commented out code. If you are planning to come back to it, the best option is to add a little explanation in the comments above the line

File: src/BMS.cpp [hunk 3]
```diff
@@ -109,15 +111,15 @@ void BMS::initializationErrorState() {
...
     }
 }
 
 void BMS::factoryInitState() {
     if (stateChanged) {
         bmsOK.writePin(BMS_NOT_OK);
         stateChanged = false;
-        clearVoltageReadings();
+        //clearVoltageReadings();
```
Comment (cbolles) [inline src/BMS.cpp:122]: NIT: Same thing here on commented out code

File: src/dev/BQ76952.cpp [hunk 4]
```diff
@@ -227,25 +227,39 @@ BQ76952::Status BQ76952::communicationStatus() {
     return BQ76952::Status::ERROR;
 }
 
-BQ76952::Status BQ76952::getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t* sum) {
+BQ76952::Status BQ76952::getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t* sum, int16_t* minVoltage, uint8_t* minCellVoltageID, int16_t* maxVoltage, uint8_t* maxCellVoltageId) {
     Status status = Status::OK;
     uint8_t cellVoltageReg = CELL_VOLTAGE_BASE_REG;
 
     uint32_t currentVoltage = 0;
+    uint16_t currentMinVoltage = 4;
```
Comment (cbolles) [inline src/dev/BQ76952.cpp:235]: This is where the code can be simplified a bit having a struct to store the values

File: targets/BMS/main.cpp [hunk 5]
```diff
@@ -172,28 +173,34 @@ int main() {
...
+    CONodeInit(&canNodeOne, &canSpecOne);
+    CONodeStart(&canNodeOne);
+    CONmtSetMode(&canNodeOne.Nmt, CO_OPERATIONAL);
 
     // Main processing loop, contains the following logic
     // 1. Update CANopen logic and processing incomming messages
     // 2. Run per-loop BMS state logic
     // 3. Wait for new data to come in
     while (1) {
+        //        uart.printf("Hello Again");
```
Comment (cbolles) [inline targets/BMS/main.cpp:194]: NIT: Same thing here on commented out code

**Review by cbolles — CHANGES_REQUESTED** [review]
Looks good, a few improvement suggestions

File: include/BMS/dev/BQ76952.hpp [hunk 6]
```diff
@@ -159,9 +159,18 @@ class BQ76952 {
...
      *            NUM_CELLS in size
      * @param[out] sum The total voltage across all cells
+     *
+     * @param[out] minVoltage Voltage value of the lowest cell voltage of all cells in the pack.
+     *
+     * @param[out] maxVoltage Voltage value of the highes cell voltage of all cells in the pack.
+     *
+     * @param[out] minCellVoltageID The ID of the cell with the lowest voltage.
+     *
+     * @param[out] maxCellVoltageId The ID of the cell with the highest voltage.
```
Comment (AarC10) [inline include/BMS/dev/BQ76952.hpp:169]: Params shouldn't be separated by newlines

**Review by AarC10 — CHANGES_REQUESTED** [review]

**Comment by mjmagee991** [comment]
[BMS TPDOs](https://trello.com/c/oSHgait7/22-bms-tpdos)

File: include/BMS/BMS.hpp [hunk 7]
```diff
@@ -365,117 +438,677 @@ class BMS {
...
+        {
+            .Key = CO_KEY(0x1A00, 4, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2100, 4, 16),
+        },
+        {
+            .Key = CO_KEY(0x1A00, 5, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2100, 5, 8),
+        },
```
Comment (mjh9585) [inline include/BMS/BMS.hpp:700]: To improve readability of the object dictionary, it would nice to include a short comment as to what each mapping entry is linked to.

File: include/BMS/BMS.hpp [hunk 8]
```diff
@@ -365,117 +438,677 @@ class BMS {
...
+        {
+            .Key = CO_KEY(0x2100, 1, CO_UNSIGNED8 | CO_OBJ___PR_),
+            .Type = nullptr,
+            .Data = (uintptr_t) &batteryVoltage,
+        },
+        {
+            .Key = CO_KEY(0x2100, 2, CO_UNSIGNED8 | CO_OBJ___PR_),
+            .Type = nullptr,
+            .Data = (uintptr_t) &minCellVoltage,
+        },
```
Comment (mjh9585) [inline include/BMS/BMS.hpp:937]: A number of these elements are defined as UNSIGNED8 but the variable they are linked to are an int16. Check the sizes and types of variables you are representing in the object dictionary.

File: include/BMS/BMS.hpp [hunk 9]
```diff
@@ -365,117 +438,677 @@ class BMS {
...
+        {
+            .Key = CO_KEY(0x1A00, 3, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2100, 3, 8),
+        },
+        {
+            .Key = CO_KEY(0x1A00, 4, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = CO_LINK(0x2100, 4, 16),
+        },
```
Comment (mjh9585) [inline include/BMS/BMS.hpp:695]: Make sure the link matches the size of the entry it is linked to. These links are defined as 16 bit, but the entry they are linked to are only 8 bit (look at other comment).

File: include/BMS/dev/BQ76952.hpp [hunk 10]
```diff
@@ -153,27 +154,24 @@ class BQ76952 {
...
+
     /**
      * Fill a buffer with the per cell voltage.
      *
-     * @param[out] cellVoltages The buffer to fill with the cell voltage. Must be
-     *            NUM_CELLS in size
+     * @param[out] cellVoltages The buffer to fill with the cell voltage. Must be NUM_CELLS in size
      * @param[out] sum The total voltage across all cells
-     *
+     * @param[out] voltageInfo A struct containing the values below.
```
Comment (mjmagee991) [inline include/BMS/dev/BQ76952.hpp:165]: This documentation needs to be updated

File: include/BMS/dev/BMSInfo.hpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+#include <stdint.h>
+
+struct cellVoltageInfo {
```
Comment (mjmagee991) [inline include/BMS/dev/BMSInfo.hpp:4]: Add a docstring explaining what this is

File: include/BMS/dev/BMSInfo.hpp [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+    int16_t minVoltage;
+    uint8_t minCellVoltageID;
+    int16_t maxVoltage;
+    uint8_t maxCellVoltageId;
+};
```
Comment (mjmagee991) [inline include/BMS/dev/BMSInfo.hpp:9]: Add a newline to the end of the file

File: CMakeLists.txt [hunk 13]
```diff
@@ -47,7 +47,7 @@ project(${BOARD_LIB_NAME}
         LANGUAGES CXX C
 )
 
-add_library(${PROJECT_NAME} STATIC)
+add_library(${PROJECT_NAME} STATIC include/BMS/dev/BMSInfo.hpp)
```
Comment (mjmagee991) [inline CMakeLists.txt:50]: Is this necessary? If so, why?

File: include/BMS/BMS.hpp [hunk 14]
```diff
@@ -258,6 +233,7 @@ class BMS {
      */
     uint16_t cellVoltage[DEV::BQ76952::NUM_CELLS];
 
+    cellVoltageInfo voltageInfo;
```
Comment (mjmagee991) [inline include/BMS/BMS.hpp:236]: Add a docstring explaining what this is

File: include/BMS/dev/BMSInfo.hpp [hunk 15]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+#include <stdint.h>
+
+struct cellVoltageInfo {
+    int16_t minVoltage;
```
Comment (mjmagee991) [inline include/BMS/dev/BMSInfo.hpp:5]:
```suggestion
    int16_t minCellVoltage;
```

File: include/BMS/dev/BMSInfo.hpp [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+    uint8_t minCellVoltageID;
+    int16_t maxVoltage;
```
Comment (mjmagee991) [inline include/BMS/dev/BMSInfo.hpp:7]:
```suggestion
    int16_t maxCellVoltage;
```

File: src/BMS.cpp [hunk 17]
```diff
@@ -135,7 +135,7 @@ void BMS::transferSettingsState() {
         bqSettingsStorage.resetTranfer();
         numAttemptsMade = 0;
         stateChanged = false;
-        clearVoltageReadings();
+        //VoltageReadings();
```
Comment (mjmagee991) [inline src/BMS.cpp:138]: Why is this commented out?

File: src/BMS.cpp [hunk 18]
```diff
@@ -261,16 +261,20 @@ bool BMS::isHealthy() {
...
 }
 
 void BMS::clearVoltageReadings() {
-    totalVoltage = 0;
-
-    // Zero out each cell voltage
-    for (uint8_t i = 0; i < DEV::BQ76952::NUM_CELLS; i++) {
-        cellVoltage[i] = 0;
-    }
+//    totalVoltage = 0;
```
Comment (mjmagee991) [inline src/BMS.cpp:272]: Why is this commented out?

File: src/dev/BQ76952.cpp [hunk 19]
```diff
@@ -226,26 +226,45 @@ BQ76952::Status BQ76952::communicationStatus() {
     }
     return BQ76952::Status::ERROR;
 }
+// Create struct that will hold CAN interrupt parameters
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:229]: Clarify or delete this comment

File: src/dev/BQ76952.cpp [hunk 20]
```diff
@@ -226,26 +226,45 @@ BQ76952::Status BQ76952::communicationStatus() {
...
 
-BQ76952::Status BQ76952::getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t* sum) {
+
+BQ76952::Status BQ76952::getCellVoltage(uint16_t cellVoltages[NUM_CELLS], uint32_t& sum, cellVoltageInfo& voltageInfo) {
     Status status = Status::OK;
     uint8_t cellVoltageReg = CELL_VOLTAGE_BASE_REG;
-
+    //Must use temporary storage variables or else the values displayed in PEAK will be inaccurate from regular changes.
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:235]:
```suggestion
    //Must use temporary storage variables or else the values reported over CAN will be inaccurate from regular changes.
```

File: src/dev/BQ76952.cpp [hunk 21]
```diff
@@ -226,26 +226,45 @@ BQ76952::Status BQ76952::communicationStatus() {
...
     uint32_t currentVoltage = 0;
+    int16_t currentMinVoltage = 4;
+    int16_t currentMaxVoltage =0 ;
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:238]: It looks like the GitHub action didn't run on your latest commit, which is why the autoformatter didn't fix this. Be sure to get it running again before pulling.

File: src/dev/BQ76952.cpp [hunk 22]
```diff
@@ -226,26 +226,45 @@ BQ76952::Status BQ76952::communicationStatus() {
...
+    uint8_t currentMinCellID;
+    uint8_t currentMaxCellID;
+
     // Loop over all the cells and update the cooresponding voltage
     for (uint8_t i = 0; i < NUM_CELLS; i++) {
+
         status = makeDirectRead(cellVoltageReg, &cellVoltages[i]);
+        status = Status::OK;
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:246]: If I remember correctly, you were using this for testing, but it should now be removed

File: targets/BMS/CMakeLists.txt [hunk 23]
```diff
@@ -2,6 +2,5 @@ include(${EVT_CORE_DIR}/cmake/evt-core_build.cmake)
 
 project(BMSmain)
 cmake_minimum_required(VERSION 3.15)
-
 make_exe(${PROJECT_NAME} main.cpp)
-target_link_libraries(${PROJECT_NAME} PUBLIC ${BOARD_LIB_NAME})
+target_link_libraries(${PROJECT_NAME} PUBLIC ${BOARD_LIB_NAME})
```
Comment (mjmagee991) [inline targets/BMS/CMakeLists.txt:6]: Add a newline to the end of the file

File: targets/BMS/main.cpp [hunk 24]
```diff
@@ -1,3 +1,6 @@
+//
+// Created by trevo on 10/6/2022.
```
Comment (mjmagee991) [inline targets/BMS/main.cpp:2]: Remove this comment

File: targets/BMS/main.cpp [hunk 25]
```diff
@@ -159,7 +158,7 @@ int main() {
     canStackDriver.Timer = &timerDriver;
     canStackDriver.Nvm = &nvmDriver;
 
-    CO_NODE_SPEC canSpec = {
+    CO_NODE_SPEC canSpecOne = {
```
Comment (mjmagee991) [inline targets/BMS/main.cpp:161]: Why did you rename this?

File: targets/CMakeLists.txt [hunk 26]
```diff
@@ -4,4 +4,4 @@ add_subdirectory(bq_interface)
 add_subdirectory(bq_settings)
 add_subdirectory(eeprom_dump)
 add_subdirectory(settings_transfer)
-add_subdirectory(system_detect)
+add_subdirectory(system_detect)
```
Comment (mjmagee991) [inline targets/CMakeLists.txt:7]: Add a newline to the end of this file

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Looks pretty good. I found a couple small things that need to be changed

File: targets/BMS/main.cpp (hunk 25 above)
Comment (tmb7852) [reply targets/BMS/main.cpp:161]: At one point the idea of this branch was to research having more than one node. I believe this change is residual from when it was handed to me. There was a canSpecOne-canSpecThree. This is no longer needed as we found that to be an unnecessary change. I will change it bac to canSpec along with the name of the node to canNode (It is currently canNodeOne).

File: src/BMS.cpp (hunk 18 above)
Comment (tmb7852) [reply src/BMS.cpp:272]: I did that for testing purposes. It has been fixed.

File: include/BMS.hpp [hunk 27]
```diff
@@ -337,21 +386,19 @@ class BMS {
...
         // 0: The TPDO number, default 0
-        // 1: The COB-ID used by TPDO0, provided as a function of the TPDO
-        //    number
-        // 2: How the TPDO is triggered, defaulting to manual trigger
+        // 1: The COB-ID used by TPDO0, provided as a function of the TPDO number
+        // 2: How the TPO is triggered, default to manual triggering
         // 3: Inhibit time, defaults to 0
-        // 5: Timer trigger time in 1ms units, 0 will disable the timer based
-        //    triggering
+        // 5: Timer trigger time in 1ms units, 0 will disable the timer based triggering
```
Comment (ActuallyZach) [inline include/BMS.hpp:392]: Should be 3, 4 not 3, 5. this seems to be an issue throughout most of the docs for the TPDOs

File: include/BMS.hpp [hunk 28]
```diff
@@ -369,113 +416,707 @@ class BMS {
...
+        {
+            .Key = CO_KEY(0x1801, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 1,
+        },
+        {
+            .Key = CO_KEY(0x1801, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0) + NODE_ID + 1,
```
Comment (ActuallyZach) [inline include/BMS.hpp:433]: Why do you add 1 to the end of this? If we are trying to access the first PDO instead of the 0 PDO shouldn't we just  use `CO_COBID_TPDO_DEFAULT(1) + NODE_ID`

File: include/BMS.hpp [hunk 29]
```diff
@@ -369,113 +416,707 @@ class BMS {
...
+        // 5: Timer trigger time in 1ms units, 0 will disable the timer based triggering
+        {
+            .Key = CO_KEY(0x1802, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 2,
+        },
+        {
+            .Key = CO_KEY(0x1802, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0) + NODE_ID + 2,
```
Comment (ActuallyZach) [inline include/BMS.hpp:465]: Same as above, why are we adding 2?

File: include/BMS.hpp [hunk 30]
```diff
@@ -369,113 +416,707 @@ class BMS {
...
+        // 5: Timer trigger time in 1ms units, 0 will disable the timer based triggering
+        {
+            .Key = CO_KEY(0x1803, 0, CO_UNSIGNED8 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) 3,
+        },
+        {
+            .Key = CO_KEY(0x1803, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
+            .Type = nullptr,
+            .Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(1) + NODE_ID,
```
Comment (ActuallyZach) [inline include/BMS.hpp:497]: Why is TPDO3 using the first TPDO Default? Shouldn't this be 3

File: targets/DEV1-BMS/main.cpp [hunk 31]
```diff
@@ -198,7 +198,7 @@ int main() {
         // Handle executing timer events that have elapsed
         COTmrProcess(&canNode.Tmr);
         // Update the state of the BMS
-        bms.process();
+        bms.canTest();
```
Comment (ActuallyZach) [inline targets/DEV1-BMS/main.cpp:201]: Should this be canTest or should it be process? It is not in the testing file so I feel like it should be process.

**Review by ActuallyZach — CHANGES_REQUESTED** [review]

File: include/BMS.hpp (hunk 28 above)
Comment (mjmagee991) [reply include/BMS.hpp:433]: This is because we're bending the rules of the CANopen to allow us to have more than just 4 TPDOs. We're basically acting like the BMS has 3 node IDs in a row

File: targets/DEV1-BMS/main.cpp (hunk 31 above)
Comment (mjmagee991) [reply targets/DEV1-BMS/main.cpp:201]: Yeah, that's my bad. Forgot to remove after testing

File: include/BMS.hpp (hunk 28 above)
Comment (ActuallyZach) [reply include/BMS.hpp:433]: Ahh that makes sense. So the additions at the end are basically to just increase the NODE_ID and give us extra TPDOs.

**Review by ActuallyZach — APPROVED** [review]

**Review by DeLozM8 — APPROVED** [review]

**Review by zsteig — APPROVED** [review]
